db/indexes_test.go

```diff
@@ -104,7 +105,99 @@ func TestInitializeIndexes(t *testing.T) {
 			}
 		})
 	}
+}
+
+// TestInitializeIndexesConcurrentMultiNode simulates a multi-node SG cluster starting up and racing to create indexes.
+func TestInitializeIndexesConcurrentMultiNode(t *testing.T) {
+	if base.TestsDisableGSI() {
+		t.Skip("This test only works with Couchbase Server and UseViews=false")
+	}
+	base.LongRunningTest(t)
+
+	base.SetUpTestLogging(t, base.LevelInfo, base.KeyAll)
+
+	// Pick a high enough number that it's likely to trigger the concurrent/race conditions we're testing for.
+	const numNodes = 10
+
+	tests := []struct {
+		numNodes         int
+		xattrs           bool
+		collections      bool
+		clusterN1QLStore bool
+	}{
+		{numNodes, true, false, false},
+		{numNodes, false, false, false},
+		{numNodes, true, true, false},
+		{numNodes, false, true, false},
+		{numNodes, true, false, true},
+		{numNodes, false, false, true},
+		{numNodes, true, true, true},
+		{numNodes, false, true, true},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("xattrs=%v collections=%v clusterN1QL=%v", test.xattrs, test.collections, test.clusterN1QLStore), func(t *testing.T) {
+			options := DatabaseContextOptions{
+				EnableXattr: test.xattrs,
+			}
+			if test.collections {
+				base.TestRequiresCollections(t)
+				// uses Scopes implicitly through test harness
+			} else {
+				options.Scopes = GetScopesOptionsDefaultCollectionOnly(t)
+			}
+			db, ctx := SetupTestDBWithOptions(t, options)
+			defer db.Close(ctx)
+			collection := GetSingleDatabaseCollection(t, db.DatabaseContext)
+
+			gocbBucket, err := base.AsGocbV2Bucket(db.Bucket)
+			require.NoError(t, err)
+
+			n1qlStore, ok := base.AsN1QLStore(collection.dataStore)
+			require.True(t, ok)
+
+			if test.clusterN1QLStore {
+				n1qlStore, err = base.NewClusterOnlyN1QLStore(gocbBucket.GetCluster(), gocbBucket.BucketName(), collection.ScopeName, collection.Name)
+				require.NoError(t, err)
+			}
+
+			// drop and restore indexes between tests, to the way the bucket pool would
```

Comment:
I'd be very tempted to move this to `db/indextest` because we know that dropping and readding indexes takes a lot of time and also causes a lot of load on the server.

There's a decent amount of reusable test code there as well, this could probably just call https://github.com/couchbase/sync_gateway/blob/main/db/indextest/indextest_test.go#L248 in a loop with different ctxs

Reply:
Done, had to add some additional stuff to `setupN1QLStore` to make it work, and it's not really compatible with the other associated test, given the permutations intentionally ignoring the test env vars for better coverage.